src/runtime.rs

```diff
@@ -0,0 +1,72 @@
+use std::any::Any;
+
+use futures::Future;
+
+pub trait RuntimeHandle: Clone + Send + Sync + 'static {
+    type JoinErr: JoinError + std::error::Error;
+    type JoinFuture<O: Send + 'static>: Future<Output = Result<O, Self::JoinErr>>
+        + Send
+        + Unpin
+        + 'static;
+    type SleepFuture: Future<Output = ()> + Send;
+
+    fn spawn<O, F>(&self, f: F) -> Self::JoinFuture<O>
+    where
+        O: Send + 'static,
+        F: Future<Output = O> + Send + 'static;
+
+    fn spawn_blocking<O, F>(&self, f: F) -> Self::JoinFuture<O>
+    where
+        O: Send + 'static,
+        F: FnOnce() -> O + Send + 'static;
+
+    fn sleep(&self, duration: std::time::Duration) -> Self::SleepFuture;
+}
+
+pub trait JoinError {
+    fn is_panic(&self) -> bool;
+    fn into_panic(self) -> Box<dyn Any + Send + 'static>;
+}
+
+#[cfg(feature = "tokio")]
+mod tokio_runtime {
+    use super::*;
+    use tokio::task::JoinError as TokioJoinError;
+
+    impl RuntimeHandle for tokio::runtime::Handle {
+        type JoinErr = tokio::task::JoinError;
+        type JoinFuture<O: Send + 'static> = tokio::task::JoinHandle<O>;
+        type SleepFuture = tokio::time::Sleep;
+
+        fn spawn<O, F>(&self, f: F) -> Self::JoinFuture<O>
+        where
+            O: Send + 'static,
+            F: Future<Output = O> + Send + 'static,
+        {
+            self.spawn(f)
+        }
+
+        fn spawn_blocking<O, F>(&self, f: F) -> Self::JoinFuture<O>
+        where
+            O: Send + 'static,
+            F: FnOnce() -> O + Send + 'static,
+        {
+            self.spawn_blocking(f)
+        }
+
+        fn sleep(&self, duration: std::time::Duration) -> Self::SleepFuture {
+            let _guard = self.enter();
+            tokio::time::sleep(duration)
+        }
+    }
```

Comment:
I think this is a good idea to explore later one, because:

1. CPU intensive nature of running the repo seems to be in favor of a native thread(or perhaps `spawn_blocking` above, although it isn't so much blocking are rather running without yielding). 
3. If one uses `spawn_blocking`, then `F` could simply be the current `repo.run()` but without the native thread, I think. So it can be explored later on. 
3. If we want `F` to be "the function that runs the repo logic", and make it async friendly yet not using any specific runtime, I think we'd have to do something quite complicated in terms of deconstructing the whole thing as a state machine that would progress when polled? In other words, more complicated then just running a select loop over a bunch of channels.

Reply:
As you know, I am a big fan of the "state machine which progresses when polled" approach and I lifted this trait straight out of such an implementation so it's possibly over complex. It's entirely possible that we could do away with this if we're using a thread internally and if that's the case then I would be happy.